### DISCUSSION TITLE
RabbitMQ instances are losing connection with worker server

### DISCUSSION THREAD

**Discussion by KPrathapReddy141** [php-amqplib/php-amqplib#1141] [Q&A]
# Subject

I have worker server which is taking messages from rabbitMQ and process those. I am facing time limit issue with the services instances connected with rabbitMQ. as I mentioned though the service instances are up and running the connection is loosing with rabbitMq.

Is there any time limit for RabbitMQ server connections hosted on AmazonMQ?

# Environment
PHP 8.2
rabbitMQ 3.11
amqp 0.9.1

# Service file

```
[Unit]
Description="Pdf_Worker instance #%i" [worker@.service](mailto:worker@.service) units
PartOf=pdf_workers.target

[Service]
Type=simple
ExecStart=/usr/bin/php /var/worker/pdf-worker/src/pdf_worker.php %i
Restart=always
```

This My target file.

```
[Unit]
Description=Pdf_Workers
Wants=[pdf_worker@1.service](mailto:pdf_worker@1.service) [pdf_worker@2.service](mailto:pdf_worker@2.service) [pdf_worker@3.service](mailto:pdf_worker@3.service)

[Install]
WantedBy=multi-user.target
```

# Consumer

```
// Establish an SSL-encrypted connection to RabbitMQ server with provided options.
    $connection = new AMQPSSLConnection($host, $port, $user, $password, $virtualHost,$sslOptions);
    $channel = $connection->channel();
    $channel->queue_declare($queue, false, true, false, false);
    // Declare a fanout exchange for message broadcasting.
    $channel->exchange_declare($exchange, 'fanout', false, true, false);
    // Bind the queue to the fanout exchange.
    $channel->queue_bind($queue, $exchange);

    $logMessage="Instance still working ..";
    FormPDFLogAndNotifyManager::pdfWorkerLoggingMethod($logMessage);

    // A callback function for processing incoming messages and log them.
    $callback = function ($msg){
        $messageData = $msg->body;

        //deserialize and extract compressed data.
        $dataArray = unserialize(gzinflate($messageData));
    };
    //Consumes messages from a specified Queue in RabbitMQ
    $channel->basic_consume($queue, $consumerTag, false, true, false, false, $callback);

    function shutdown($channel, $connection)
    {
        $channel->close();
        $connection->close();
    }
    //This function executes a call back function shutdown when the connection about to shut down due to error or any other reasons.
    register_shutdown_function('shutdown', $channel, $connection);

    // Loop as long as the channel has callbacks registered
    $channel->consume();

    FormPDFLogAndNotifyManager::pdfWorkerLoggingMethod('end of worker.');

} catch (Exception $e) {
    $errorMessage='AmazonMQ Error Message ::' . $e->getMessage();
    FormPDFLogAndNotifyManager::pdfWorkerLoggingMethod($errorMessage);
```

**Comment by lukebakken** [comment]
Hello, you _MUST_ provide the RabbitMQ log files and let us know when your service lost connection so we can see what RabbitMQ logged at the same time.

More than likely your application is taking too long to process a message.

**Comment by KPrathapReddy141** [comment]
Hello,

Thanks for the response.

In Amazon MQ, I may not have direct access to the RabbitMQ configuration files since Amazon MQ is a managed service, and configuration options are typically managed through the AWS Management Console or the AWS CLI.

In the consumer file we can see that in some points I am trying to log the message if instance of the file is created(11 and 12 lines). and also writing when it close the connection(5th line from down).

```
2023-11-14 16:01:40 Instance #2 still working ..
2023-11-14 16:01:40 Instance #3 still working ..
2023-11-14 16:01:40 Instance #1 still working ..
```

As you can see, I can only see these lines has written after losing the connection it didn't write **end of worker** for any of the instance. can you suggest is there any other way option to take logs?

> More than likely your application is taking too long to process a message.

I don't know why you said that. can you elaborate why it's gonna take too much time to process the message.

**Reply by lukebakken** [reply]
> > More than likely your application is taking too long to process a message.

> I don't know why you said that. can you elaborate why it's gonna take too much time to process the message.

Since we don't have RabbitMQ log files, all I can do is guess.

My **guess** is that your application is taking too long to process a message and is either blocking heartbeats, or exceeding RabbitMQ's consumer timeout (https://www.rabbitmq.com/consumers.html#acknowledgement-timeout).

If we had RabbitMQ log files we would know very quickly.

**Reply by KPrathapReddy141** [reply]
I don't think this is happening in this scenario. what I observed if consumer file instances are up and running and made connection with the RabbitMQ if there is no message to process it is losing the connection with rabbitMq 3 to 5 min. if there are messages to process are continuous publishing the connection stays Langer then  this let's say 20 min but eventually after some point of the losing the connection with RabbitMQ though the consumer file instances are in active state.

with this info have any idea why it's loosing connection when there are no messages to process for certain amount of time like 3 to 5.

One more thing to mention when it is lost the connection with consumer instances, the ranbbitMq instances won't disappear from AmazonMq web console. it will stay there but it doesn't have any connection and not able to process the message when exchange assign a message. ranbbitMq instance takes the message try to process it but it doesn't have connection with consumer so it disappear with the message.

**Reply by lukebakken** [reply]
There is nothing we can do to assist you without one or both of these:
* RabbitMQ logs
* Code that we can run to reproduce the issue

**Reply by KPrathapReddy141** [reply]
Hey Luke,

I can provide the code for both producer and consumer files.

Producer:


```
$connection = new AMQPSSLConnection($host, $port, $user, $password, $virtualHost, $sslOptions);
            $channel = $connection->channel();
            //serialize and compress data if not a basic data value
            $value = serialize([$view,$specific_form_id,$specific_form,$form_pdf_id,$bucket,$file_path,$notify_end_point]); //needed to flatten complex data structures such as arrays and class instances
            $value = gzdeflate($value, 5); //compress data

            // Declare a fanout exchange for message broadcasting.
            $channel->exchange_declare($exchangeName, $exchangeType, false, true, false);
            //$channel->queue_declare($queueName, false, true, false, false);
            $messageBody = $value;
            $message = new AMQPMessage($messageBody);
            //publishing the message to the specified queue and exchange.
            $channel->basic_publish($message, $exchangeName);
```

Consumer:

```
// Establish an SSL-encrypted connection to RabbitMQ server with provided options.
    $connection = new AMQPSSLConnection($host, $port, $user, $password, $virtualHost,$sslOptions);
    $channel = $connection->channel();
    $channel->queue_declare($queue, false, true, false, false);
    // Declare a fanout exchange for message broadcasting.
    $channel->exchange_declare($exchange, 'fanout', false, true, false);
    // Bind the queue to the fanout exchange.
    $channel->queue_bind($queue, $exchange);

    $logMessage="Instance still working ..";
    FormPDFLogAndNotifyManager::pdfWorkerLoggingMethod($logMessage);

    // A callback function for processing incoming messages and log them.
    $callback = function ($msg){
        $messageData = $msg->body;

        //deserialize and extract compressed data.
        $dataArray = unserialize(gzinflate($messageData));
    };
    //Consumes messages from a specified Queue in RabbitMQ
    $channel->basic_consume($queue, $consumerTag, false, true, false, false, $callback);

    function shutdown($channel, $connection)
    {
        $channel->close();
        $connection->close();
    }
    //This function executes a call back function shutdown when the connection about to shut down due to error or any other reasons.
    register_shutdown_function('shutdown', $channel, $connection);

    // Loop as long as the channel has callbacks registered
    $channel->consume();

    FormPDFLogAndNotifyManager::pdfWorkerLoggingMethod('end of worker.');

} catch (Exception $e) {
    $errorMessage='AmazonMQ Error Message ::' . $e->getMessage();
    FormPDFLogAndNotifyManager::pdfWorkerLoggingMethod($errorMessage);
}
```

**Reply by lukebakken** [reply]
Is that your **COMPLETE** consumer code or is it part of a larger application?

In addition, **you must contact Amazon support to get the RabbitMQ logs**.

**Reply by KPrathapReddy141** [reply]
That is my consumer code, it is not part of any large file.
I will do that and try to get the log file.

**Reply by KPrathapReddy141** [reply]
Hey Luke,

I am thinking of using timer file to restart the consumer instances for every 3 minutes and reactivate connection with RabbitmQ.

```
[Unit]
Description=Restart Pdf_Workers targets every 3 minutes

[Timer]
OnCalendar=*:0/3
Persistent=true

[Install]
WantedBy=timers.target
```

this file triggers all the service instances for every 3 minutes. do you think is this a best practice to do this or should I look for other options. and this is persistent so, there is no chance to loose the message when a consumer instance goes down after consume the message and try to process. systemMd persist the message and will catch up with the triggered messages before restart.
